### PR TITLE
[language] LibraVM and MoveVm initial separation

### PR DESCRIPTION
**Pull request by dariorussi**
This is the initial break up of the Libra execution logic in a LibraVM (name to be finalized) and
a MoveVM.
The separation allows for the 2 layers to move independently and it also cleans up a lot of lifetime
parameters that are not needed any longer.
This also changes the outside type used: it was `MoveVM` and it's changed to `LibraVM`.
The latter will likely change name but I wanted to free the MoveVM name to be used for what it
really is. That is a mechanical change so it should be fine.

The entry points are in `move_vm.rs` (as it used to) and that module has got more code from other files.
All the entry points into the interpreter have been moved to `VMRuntime` and `TransactionExecutor`
holds much less state than it used to be. `TransactionExecutor` essentially has some 
transaction execution logic and all execution is forwarded to `VMRuntime`.

The state machine for the transaction is still around as it used to be, however bytecode verification has
been moved to the execution state. That is in line with the new behavior and something that may evolve still.

There is another clean up that should happen soon and it's the `StateView` and `VMRuntime` 
relationship which is something that should be changed by removing the needed `StateView` 
and use the data view instead.
In other words, the VM should use the data view (a concept of that) to fetch code.

Overall this code will evolve dramatically in the next 2 weeks so I don't think there is much value
in cleaning it up any more. Evolutions of the Libra layer and the Move layer will likely rewrite a
significant portion of this code base so let's keep "change requests" to a minimum.

### CODE REVIEW

File: language/vm/vm-runtime/src/execution_context.rs [hunk 1]
```diff
@@ -94,6 +90,8 @@ pub trait InterpreterContext {
     fn deduct_gas(&mut self, amount: GasUnits<GasCarrier>) -> VMResult<()>;
 
     fn remaining_gas(&self) -> GasUnits<GasCarrier>;
+
+    fn exists_module(&self, m: &ModuleId) -> bool;
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/execution_context.rs]: it's convenient to add this here for now. The "conversation" between the Libra and VM layer is evolving and this may go elsewhere....

File: language/vm/vm-runtime/src/move_vm.rs [hunk 2]
```diff
@@ -72,17 +90,148 @@ impl VMExecutor for MoveVM {
...
+                // TODO: Implement the logic for processing writeset transactions.
+                TransactionBlock::WriteSet(_) => unimplemented!(""),
+            }
+        }
+
+        Ok(result)
+    }
+}
+
+/// Transactions divided by transaction flow.
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/move_vm.rs]: I moved this block logic here because it's the entry point.
This is likely to evolve with the new design of the Libra layer (@runtian-zhou) so I am just having this close to the entry point and for simplicity

File: language/vm/vm-runtime/src/move_vm.rs [hunk 3]
```diff
@@ -72,17 +90,148 @@ impl VMExecutor for MoveVM {
...
+/// 2. The script to be executed is in the whitelist.
+/// 3. Invokes `LibraAccount.prologue`, which checks properties such as the transaction has the
+/// right sequence number and the sender has enough balance to pay for the gas. 4.
+/// Transaction arguments matches the main function's type signature. 5. Script and modules
+/// in the transaction pass the bytecode static verifier.
+///
+/// Note: In the future. we may defer these checks to a later pass, as all the scripts we will
+/// execute are pre-verified scripts. And bytecode verification is expensive. Thus whether we
+/// want to perform this check here remains unknown.
+pub fn verify_transaction(
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/move_vm.rs]: this is only public because of testing but it seems we need a better way to plug in tests and understand the dependency graph

File: language/vm/vm-runtime/src/move_vm.rs [hunk 4]
```diff
@@ -72,17 +90,148 @@ impl VMExecutor for MoveVM {
...
+
+/// Transactions divided by transaction flow.
+/// Transaction flows are different across different types of transactions.
+pub enum TransactionBlock {
+    UserTransaction(Vec<SignedTransaction>),
+    WriteSet(WriteSet),
+    BlockPrologue(BlockMetadata),
+}
+
+pub fn chunk_block_transactions(txns: Vec<Transaction>) -> Vec<TransactionBlock> {
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/move_vm.rs]: this is only public because of testing but it seems we need a better way to plug in tests and understand the dependency graph

File: language/vm/vm-runtime/src/move_vm.rs (hunk 3 above)
Comment (dariorussi) [inline language/vm/vm-runtime/src/move_vm.rs]: this is the validation story so I moved it here because it's the entry point.
Could live elsewhere and it probably will

File: language/vm/vm-runtime/src/process_txn/execute.rs [hunk 5]
```diff
@@ -66,57 +60,44 @@ where
...
+                        warn!("[VM] User error running script: {:?}", err);
+                        return executor.failed_transaction_cleanup(runtime, state_view, Err(err));
+                    }
+                },
             };
-            let module_bytes = module.into_inner();
-            let output = txn_executor.transaction_cleanup(vec![(module_id, module_bytes)]);
+
+            executor.transaction_cleanup(runtime, state_view, vec![(module_id, module)])
+            // REVIEW: should this go in ever? essentially should a module be published eagerly?
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/process_txn/execute.rs]: I'd like to leave this here as is and think some more about it.
I am not worried about this sticking in the code because this logic is going to be sorted out next

File: language/vm/vm-runtime/vm-runtime-types/src/value.rs [hunk 6]
```diff
@@ -362,6 +363,16 @@ impl Value {
     pub fn pretty_string(&self) -> String {
         self.0.pretty_string()
     }
+
+    pub fn is_valid_script_arg(&self, sig: &SignatureToken) -> bool {
```
Comment (dariorussi) [inline language/vm/vm-runtime/vm-runtime-types/src/value.rs]: this is really a hack but again part of the more logical clean up of the Libra and the Move layer

File: language/vm/vm-runtime/src/runtime.rs [hunk 7]
```diff
@@ -52,147 +60,165 @@ impl<'alloc> VMRuntime<'alloc> {
...
-    /// 3. Invokes `LibraAccount.prologue`, which checks properties such as the transaction has the
-    /// right sequence number and the sender has enough balance to pay for the gas. 4.
-    /// Transaction arguments matches the main function's type signature. 5. Script and modules
-    /// in the transaction pass the bytecode static verifier.
-    ///
-    /// Note: In the future. we may defer these checks to a later pass, as all the scripts we will
-    /// execute are pre-verified scripts. And bytecode verification is expensive. Thus whether we
-    /// want to perform this check here remains unknown.
-    pub fn verify_transaction(
+    pub(crate) fn load_gas_schedule(
```
Comment (runtian-zhou) [inline language/vm/vm-runtime/src/runtime.rs]: Let's chat about the gas schedule later. It seems to me that loading the gas schedule should not rely on the code cache. Currently we are only using code cache to derive the access path. We may hard code this access path in the future so that loading the gas schedule will have nothing to do with the code cache.

File: language/vm/vm-runtime/src/runtime.rs [hunk 8]
```diff
@@ -52,147 +60,165 @@ impl<'alloc> VMRuntime<'alloc> {
...
+        state_view: &dyn StateView,
+        context: &mut dyn InterpreterContext,
+        txn_data: &TransactionMetadata,
+        gas_schedule: &CostTable,
+        script: Vec<u8>,
+        args: Vec<Value>,
+    ) -> VMResult<()> {
+        let main = self.script_cache.cache_script(&script)?;
+
+        if !verify_actuals(main.signature(), &args) {
```
Comment (runtian-zhou) [inline language/vm/vm-runtime/src/runtime.rs]: Love this!

File: language/vm/vm-runtime/src/runtime.rs [hunk 9]
```diff
@@ -52,147 +60,165 @@ impl<'alloc> VMRuntime<'alloc> {
...
+            &code_cache,
+            txn_data,
+            gas_schedule,
+            module,
+            function_name,
+            args,
+        )
+    }
+
+    pub fn cache_module(&mut self, module: VerifiedModule) {
```
Comment (runtian-zhou) [inline language/vm/vm-runtime/src/runtime.rs]: Should the two functions be private?

File: language/vm/vm-runtime/src/system_txn/block_metadata_processor.rs [hunk 10]
```diff
@@ -46,16 +45,21 @@ where
...
+    let mut txn_executor = TransactionExecutor::new(&gas_schedule, data_cache, txn_data);
     let result = if let Ok((id, timestamp, previous_vote, proposer)) = block_metadata.into_inner() {
         let args = vec![
             Value::u64(timestamp),
             Value::byte_array(id),
             Value::byte_array(previous_vote),
             Value::address(proposer),
         ];
-        txn_executor.execute_function(&LIBRA_SYSTEM_MODULE, &BLOCK_PROLOGUE, args)
+        txn_executor.execute_function(
```
Comment (runtian-zhou) [inline language/vm/vm-runtime/src/system_txn/block_metadata_processor.rs]: Can we use `runtime` to execute the function rather than using `TransactionExecutor`?

File: language/vm/vm-runtime/src/move_vm.rs (hunk 2 above)
Comment (runtian-zhou) [reply language/vm/vm-runtime/src/move_vm.rs]: Shall we rename this file to `libra_vm.rs`?

File: language/vm/vm-runtime/src/process_txn/execute.rs (hunk 5 above)
Comment (runtian-zhou) [reply language/vm/vm-runtime/src/process_txn/execute.rs]: IMHO with our current setup, we should uncomment this block of code right now as the code cache is only hooked up with StateView rather than the BlockDataCache. Thus we have to publish it eagerly so that later transaction can see this new module.

File: language/vm/vm-runtime/src/runtime.rs (hunk 7 above)
Comment (dariorussi) [reply language/vm/vm-runtime/src/runtime.rs]: > Let's chat about the gas schedule later. It seems to me that loading the gas schedule should not rely on the code cache. Currently we are only using code cache to derive the access path. We may hard code this access path in the future so that loading the gas schedule will have nothing to do with the code cache.

I would agree with you on this. I am not sure what the best way to do this is but it bothered me too that we had to rely in a strange way to the runtime. Think about this a bit as well

File: language/vm/vm-runtime/src/move_vm.rs (hunk 2 above)
Comment (dariorussi) [reply language/vm/vm-runtime/src/move_vm.rs]: I am ok doing it and if you think it's the right thing to do I'll do it

File: language/vm/vm-runtime/src/process_txn/execute.rs (hunk 5 above)
Comment (dariorussi) [reply language/vm/vm-runtime/src/process_txn/execute.rs]: isn't the case that later transaction will ask the data view for the new code and fetch it that way?
I mean this is publishing the code which may be fine but I wonder if we want it this way.
It seems to me at the end of a publish transaction we publish the code explicitly in the write set. Isn't that the case?

File: language/vm/vm-runtime/src/runtime.rs (hunk 9 above)
Comment (dariorussi) [reply language/vm/vm-runtime/src/runtime.rs]: I'll review that and change all functions in runtime as needed. 
We have more public functions that we'd like to because we are in the middle of this refactoring so something may be here temporarily

File: language/vm/vm-runtime/src/system_txn/block_metadata_processor.rs (hunk 10 above)
Comment (dariorussi) [reply language/vm/vm-runtime/src/system_txn/block_metadata_processor.rs]: let me look into that and I see where you are coming from and good point!

File: language/vm/vm-runtime/src/process_txn/validate.rs [hunk 11]
```diff
@@ -299,32 +273,6 @@ where
...
-    fn new(
-        metadata: TransactionMetadata,
-        gas_schedule: &'txn CostTable,
-        module_cache: P,
-        data_cache: &'txn dyn RemoteCache,
-    ) -> Self {
-        let txn_executor =
-            TransactionExecutor::new(module_cache, gas_schedule, data_cache, metadata);
-        Self { txn_executor }
+        Ok(())
```
Comment (tzakian) [inline language/vm/vm-runtime/src/process_txn/validate.rs]: Super minor nit: Could you move this `Ok(())` up into the match arm?

File: language/vm/vm-runtime/src/runtime.rs (hunk 7 above)
Comment (sblackshear) [reply language/vm/vm-runtime/src/runtime.rs]: Add a TODO for this?

File: language/vm/vm-runtime/src/txn_executor.rs [hunk 12]
```diff
@@ -162,16 +147,99 @@ where
...
+        runtime: &VMRuntime,
+        state_view: &dyn StateView,
+        address: AccountAddress,
+        module: &ModuleId,
+        function_name: &IdentStr,
+        args: Vec<Value>,
+    ) -> VMResult<()> {
+        let old_sender = self.txn_data.sender;
+        self.txn_data.sender = address;
+        let res = self.execute_function(runtime, state_view, module, function_name, args);
```
Comment (sblackshear) [inline language/vm/vm-runtime/src/txn_executor.rs]: @runtian-zhou had a good suggestion in the past that I forgot to implement: we should typecheck the args using something like the ` if !verify_actuals(main.signature(), &args)` code in `execute_script` to avoid mistakes in generating genesis. Would you mind adding a TODO for this?

File: language/vm/vm-runtime/src/system_txn/block_metadata_processor.rs (hunk 10 above)
Comment (dariorussi) [reply language/vm/vm-runtime/src/system_txn/block_metadata_processor.rs]: I looked at this because I think it's a really good idea. However it makes changes to the way the write set is computed (it lives in `TransactionExecutor` now) so I am leaving it to you.
I am not sure how you are going to reorganize that code and you are the best one to make that call

File: language/vm/vm-runtime/src/process_txn/execute.rs (hunk 5 above)
Comment (dariorussi) [reply language/vm/vm-runtime/src/process_txn/execute.rs]: ok if I understand I'll put up a TODO here and remove the code

File: language/vm/vm-runtime/src/runtime.rs (hunk 7 above)
Comment (dariorussi) [reply language/vm/vm-runtime/src/runtime.rs]: I don't think we need a TODO here, this code will continue to be changes in the next days and that will drive this refactor the proper way

File: language/vm/vm-runtime/src/runtime.rs (hunk 9 above)
Comment (dariorussi) [reply language/vm/vm-runtime/src/runtime.rs]: this is called from genesis for now and I think we'll rework that part so one of us will do it later

File: language/vm/vm-runtime/src/move_vm.rs [hunk 13]
```diff
@@ -26,42 +40,46 @@ rental! {
...
         }
     }
 }
 
-impl VMExecutor for MoveVM {
+// Executor external API
+impl VMExecutor for LibraVM {
+    /// Execute a block of transactions. The output vector will have the exact same length as the
+    /// input vector. The discarded transactions will be marked as `TransactionStatus::Discard` and
+    /// have an empty writeset. Also the data view is immutable, and also does not have interior
```
[the remark this answers is not in the document]
Comment (dariorussi) [reply language/vm/vm-runtime/src/move_vm.rs]: this is lifted from another file, it's untouched, I can review the comment a bit of course

File: language/vm/vm-runtime/src/move_vm.rs [hunk 14]
```diff
@@ -72,17 +90,148 @@ impl VMExecutor for MoveVM {
...
+pub fn verify_transaction(
+    runtime: &VMRuntime,
+    txn: SignedTransaction,
+    state_view: &dyn StateView,
+) -> Option<VMStatus> {
+    trace!("[VM] Verify transaction: {:?}", txn);
+    let mut data_cache = BlockDataCache::new(state_view);
+
+    // If we fail to load the gas schedule, then we fail to process the block.
+    let gas_schedule = match runtime.load_gas_schedule(&mut data_cache, state_view) {
```
[the remark this answers is not in the document]
Comment (dariorussi) [reply language/vm/vm-runtime/src/move_vm.rs]: I'll let @tzakian comment on that. I think you are correct that the transaction flow needs a bit of a review. But that work is coming so I think we should keep those issue in mind. I am not sure there is a TODO to add here, let me know otherwise

File: language/vm/vm-runtime/src/move_vm.rs [hunk 15]
```diff
@@ -72,17 +90,148 @@ impl VMExecutor for MoveVM {
...
+/// want to perform this check here remains unknown.
+pub fn verify_transaction(
+    runtime: &VMRuntime,
+    txn: SignedTransaction,
+    state_view: &dyn StateView,
+) -> Option<VMStatus> {
+    trace!("[VM] Verify transaction: {:?}", txn);
+    let mut data_cache = BlockDataCache::new(state_view);
+
+    // If we fail to load the gas schedule, then we fail to process the block.
```
[the remark this answers is not in the document]
Comment (dariorussi) [reply language/vm/vm-runtime/src/move_vm.rs]: removed

File: language/vm/vm-runtime/src/txn_executor.rs (hunk 12 above)
Comment (dariorussi) [reply language/vm/vm-runtime/src/txn_executor.rs]: `execute_function` verifies the arguments not sure if we want the TODO, do we?

File: language/vm/vm-runtime/src/move_vm.rs (hunk 2 above)
Comment (runtian-zhou) [reply language/vm/vm-runtime/src/move_vm.rs]: Go for it!
